### PR TITLE
Merge prototypes and setup code structure

### PR DESCRIPTION
**Pull request by karllark**
Addresses issues #2, #20.

This merges the prototypes created by @karllark and @thomaslai78226 for the v2.0 version of PAHFIT.  Along with various updates to provide a flexible base for the PAHFIT variants.  Additional changes will likely be needed, but hopefully it will provide a good start for the full v2.0 development.

Details:
1) The PAHFITBase class provides a generic PAHFIT allowing for a variable number of the classic components.  Continuum: stellar and dust emission blackbodies, dust features: Drude profiles, atomic ionized and molecular features: Gaussian profiles, and the Smith & Draine extinction curve.

2) The Classic variant is created by instantiating the PAHFITBase class with dictionaries giving the parameters of the components.  This should make it easy to add or subtract subcomponents (features, blackbodies, ...) from the components.  Using dictionaries to pass the info should also allow for optional inputs (e.g., constraints on parameters not required).  And dictionaries mean the ordering of the inputs is not important.

3) Models for the subcomponents are in a separate file and are named similarly to the astropy.modeling nomencature (e.g., Drude1D) where possible.

4) A bare bones plotting member function of PAHFITBase exists.  It uses the ability to loop through the subcomponents of a compound model to make a plot like that in PAHFIT v1.2.  At this point, it keys off the type of model, but probably we should come up with a naming convention for the subcomponent models to make this more flexible (e.g., ion_'name', h2_'name', bb_'name', df_'name') where the name identifies the specific feature.

5) Parameter names are not very useful on the full compound model (e.g., lots of 'mean_x' where x is an integer).  The parameter names on the subcomponents should make sense when used in conjunction with an appropriate subcomponent name (see point 4 above).

6) The width parameters for the subcomponents are all specified as fwhm values for consistency with astropy.modeling parameters (e.g., Gaussian1D).  Note that PAHFIT v1.2 has them as frac_fwhm.  A simple translation, but useful for making the code more generic I think.

7) There may be an issue with the extinction curve used.  There is an odd oscillation on the red side of the 10 micron silicate feature.  Not sure if this oscillation was in PAHFIT v1.2.  I have not been able to confirm that the code for this extinction curve I merged from Thomas' prototype matches the v1.2 version (not trivial to compare to my eyes).

8) A start on how the instrument and science packs could be implemented in code is included.  Each are treated as classes in the code allowing for easier coding in other areas.  Basically, the classes provide specific variables and functions that are assumed to exist by other parts of the code (e.g., fwhm in the instrument pack class).  To define a new instrument or science pack, a class created following those guidelines will transparently be used by the rest of the code.   Future enhancement would be to have the ability to read instrument/science pack data from disk (separate from this pull request).

Example 1 (nucleus of M101):
screenshot from 2018-10-24 13-32-03

Example 2 (galaxy from Thomas)
screenshot from 2018-10-24 13-32-30

### CODE REVIEW

**Comment by karllark** [comment]
Ready for review.  More work may (likely) needed, but we should start the review and see what is needed.

**Comment by karllark** [comment]
No documentation or tests yet.  Maybe they should be added here, but I'm feeling like a separate efforts/pull requests would be better.

**Comment by karllark** [comment]
Tests failing as we still do not have any - so as expected.

**Comment by karllark** [comment]
Ping.

No deadline, but the sooner we get the "base" of PAHFIT v2.0 into the master, the sooner others can start directly contributing all the needed v2.0 code, docs, tests, etc.

**Comment by jdtsmith** [comment]
Sorry this week has been a nonstop circus.  Would it work to schedule a brief call to go over it?  Thanks for working on this.

JD

**Comment by karllark** [comment]
If a brief call would help, I'm all for it.  I am busy all next week, so it would have to be today or after next week.

**Comment by els1** [comment]
Sorry. I’m only free at 5pm today and next week I’m on travel.

Best regards,
Els

---------------------------------------------------------------------------------
Els Peeters | Associate Professor
Dept. of Physics & Astronomy, University of Western Ontario
SETI Institute
Tel: +1 519 661 2111 x80973
e-mail:epeeters@uwo.ca<mailto:epeeters@uwo.ca>, epeeters@seti.org<mailto:epeeters@seti.org>
www:http://www.astro.uwo.ca/~epeeters
---------------------------------------------------------------------------------


On Nov 2, 2018, at 10:49 AM, Karl Gordon <notifications@github.com<mailto:notifications@github.com>> wrote:


If a brief call would help, I'm all for it. I am busy all next week, so it would have to be today or after next week.

**Comment by karllark** [comment]
I can do 5 pm today if that works for others.

**Comment by jdtsmith** [comment]
I'd definitely prefer more identifiable names for the components of the Drude profiles: `lam_0`, `inten` (or `central_inten`), `fwhm` (but we have to decide whether `Drude1D` will include fractional FWHM or full FWHM; fractional would be consistent with the original model).

**Comment by karllark** [comment]
I was thinking that making the names more identifiable could happen in later pull requests.  Not 100% clear how important this is given that the individual models in the component model are identifiable.  But definitely need to have unique and hopefully easily understandable names for the input/output.

fwhm or frac_fwhm is a good question.  I picked fwhm to be consistent with what is output.  If we want what is input to the same as output (#13), then it seems that it would be cleaner to stick with fwhm for the internals to the code.  This does break with the IDL version.

I had seen this pull request as mainly getting the basic code and structure in place, and then we can refine later.  But we can go a different route if that is what is needed.

**Comment by ThomasSYLai** [comment]
5 pm today is fine for me. But I need to leave before 6 pm. -Thomas

**Comment by karllark** [comment]
@jdtsmith : would 5 pm today work for you?

**Comment by karllark** [comment]
Lets go for it.  Should I setup a webex - or is there an alternative we should use?

**Comment by ThomasSYLai** [comment]
@karllark , can you help us set it up please? I normally use ZOOM, but it seems that for a group conference it will cut off by itself after certain time if only using their free service.

**Comment by els1** [comment]
Yes please.

Best regards,
Els

---------------------------------------------------------------------------------
Els Peeters | Associate Professor
Dept. of Physics & Astronomy, University of Western Ontario
SETI Institute
Tel: +1 519 661 2111 x80973
e-mail:epeeters@uwo.ca<mailto:epeeters@uwo.ca>, epeeters@seti.org<mailto:epeeters@seti.org>
www:http://www.astro.uwo.ca/~epeeters
---------------------------------------------------------------------------------


On Nov 2, 2018, at 3:27 PM, Karl Gordon <notifications@github.com<mailto:notifications@github.com>> wrote:


Lets go for it. Should I setup a webex - or is there an alternative we should use?

**Comment by karllark** [comment]
I just sent around the bluejeans link.  Same one as we used for the 1/2 day PAHFIT workshop.

File: pahfit/PAHFITBase.py [hunk 1]
```diff
@@ -175,11 +168,62 @@ def __init__(self,
...
+        cont_model = cont_components[0]
+        for cmodel in cont_components[1:]:
+            cont_model += cmodel
+        cont_y = cont_model(x)
+
+        # now plot the dust and gas lines
+        for cmodel in model:
+            if isinstance(cmodel, Gaussian1D):
+                ax.plot(x, (cont_y + cmodel(x))*ext_model/x,
+                        color='tab:purple')
```
[the remark this answers is not in the document]
Comment (karllark) [reply pahfit/PAHFITBase.py]: Nope.  Just trying to reproduce the classic PAHFIT colors and this was the "easy" way.  If it causes ipython to have issues, we should switch to another way of specifying.  Happy if you or someone else can do this.  :-)  Maybe you can open an issue on this?

**Comment by karllark** [comment]
With three reviewers happy (enough) with this code, I'll go ahead and merge tomorrow (Friday) unless someone objects before then.

After this merge, I look forward to seeing pull requests improving and fixing the code!  Remember to comment on an issues or create a new issues when you start working on something so others know.

**Comment by karllark** [comment]
@jdtsmith : I am going to leave PAHFIT.py as the name of the script.  Mainly as with use of command line arguments, other instrument and science packs can be supported.  You are right, it is an example at this point, mainly as Spitzer+Exgal is the only example we have.  :-)

**Comment by jdtsmith** [comment]
I mean I guess it looks to me like an example script that an end-user of PAHFIT would have, not a developer.  Perhaps it should eventually live in an "examples" directory where we can collect some.  This will probably sort itself out as we get more clarity on user interaction (input packs vs. command line arguments vs. optional keyword arguments, etc.).

**Pull request merged**